helm-chart/images/binderhub/Dockerfile

```diff
@@ -1,48 +1,57 @@
-# Using multi-stage builds
-ARG DIST=buster
-FROM buildpack-deps:$DIST as build-stage
-# ARG has to occur twice to be used in both contents and FROM
+# We use a build stage to package binderhub and pycurl into a wheel which we
+# then install by itself in the final image which is relatively slimmed.
 ARG DIST=buster
 
+
+# The build stage
+# ---------------
+FROM python:3.8-$DIST as build-stage
+# ARG DIST has to be defined again to made available inside this build stage.
+# ref: https://docs.docker.com/engine/reference/builder/#understand-how-arg-and-from-interact
+ARG DIST
+
+# Install node as required to package binderhub to a wheel
 RUN echo "deb http://deb.nodesource.com/node_14.x $DIST main" > /etc/apt/sources.list.d/nodesource.list \
  && curl -s https://deb.nodesource.com/gpgkey/nodesource.gpg.key | apt-key add -
-
 RUN apt-get update && \
     apt-get install --yes \
-        nodejs \
-        python3 \
-        python3-pip \
-        python3-wheel \
-        python3-setuptools
+        nodejs
 
 # Copy the whole git repository to /tmp/binderhub
 COPY . /tmp/binderhub
 WORKDIR /tmp/binderhub
 
-# Build binderhub the python library
-RUN python3 setup.py bdist_wheel
+# Build the binderhub python library into a wheel and save it to the ./dist
+# folder. Also build pycurl into a wheel as doing it requires curl-config and
+# gcc that aren't available in the slim version of the final stage.
+RUN python setup.py bdist_wheel
+RUN pip wheel pycurl --wheel-dir ./dist
+
 
 # The final stage
 # ---------------
-FROM python:3.8-$DIST
+FROM python:3.8-slim-$DIST
 WORKDIR /
 
-# Copy the built binderhub python wheel from the build-stage
-# to the current directory within the container.
-COPY --from=build-stage /tmp/binderhub/dist/*.whl .
+# The slim version doesn't include git as required by binderhub
+RUN apt-get update && \
+    apt-get install --yes \
+        git
```

Comment:
You will also have to delete the caches and other temporary files again. Otherwise the image will contain extra files.

Reply:
Thanks! I mistook the comment below from Dockerfile's best practices that we may not need to do the `rm -rf /var/lib/apt/lists/*` but it only meant we don't need to do `apt-get clean` in debian/docker base images.

> Official Debian and Ubuntu images automatically run apt-get clean, so explicit invocation is not required.